File: leetpattern/cpp/0001_0300/0134_gas_station.cc
#include <cassert>
#include <vector>
using namespace std;

class Solution {
   public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int totalTank = 0, currTank = 0;
        int start = 0;
        int n = gas.size();

        for (int i = 0; i < n; i++) {
            int diff = gas[i] - cost[i];
            totalTank += diff;
            currTank += diff;

            if (currTank < 0) {
                start = i + 1;
                currTank = 0;
            }
        }

        return (totalTank >= 0) ? start : -1;
    }
};

int main() {
    Solution obj;
    vector<int> gas{1, 2, 3, 4, 5};
    vector<int> cost{3, 4, 5, 1, 2};
    int res = obj.canCompleteCircuit(gas, cost);
    assert(res == 3);
    return 0;
}

File: leetpattern/python/0001_0300/0134_gas_station.py
from typing import List


class canCompleteCircuit:
    def greedy(self, gas: List[int], cost: List[int]) -> int:
        """
        Approach: Greedy
        Time Complexity: O(n)
        Space Complexity: O(1)
        """

        # edge case
        if sum(gas) < sum(cost):
            return -1

        cur_sum = 0
        start = 0

        for i in range(len(gas)):
            cur_sum += gas[i] - cost[i]

            if cur_sum < 0:
                start = i + 1
                cur_sum = 0

        return start


if __name__ == "__main__":
    solution = canCompleteCircuit()
    gas = [1, 2, 3, 4, 5]
    cost = [3, 4, 5, 1, 2]
    assert solution.greedy(gas, cost) == 3
